### PR TITLE
Improve linux performance when listening on many ports

### PR DESCRIPTION
**Pull request by LukeButters**
# Background

While trying to get 1000s of listening simulated tentacles working, I was running into issues in which Octopus was unable to communicate with the Tentacle. Eventually I discovered that the CPU usage of Tentacle Simulator was high despite it not doing anything. Turns out it was spending all its time in `SpinWait`.

This was done because `TcpListener.Stop hangs if Accept is in progress` https://github.com/dotnet/runtime/issues/24513 . Which is claimed to be fixed, although I see an absolute disaster of reverts so it is not clear if it is fixed.

Working on the assumption `TcpListener.Stop` still hangs if Accept is in progress, the new optional path is to:
* await waiting to accept a connection with the cancellation token.
* when the `SecureListener` is disposed first the CancellationToken is cancelled, then we wait for the accept task to complete (which will be some sort cancellation exception), after that when we know we are not in the middle of Accept we then call `TcpListener.Stop`.

Note that the new behaviour must be opted in by setting:
```
UseAsyncListener = true;
```
on `HalibutTimeoutsAndLimits`.

# Results

## Before

CPU usage was half my cores when listening on 1000 sockets.

## After

CPU usage is near nothing.

Tentacle Simulator can actually be used with 1000s of listening tentacles.

# How to review this PR

Quality :heavy_check_mark:

# Pre-requisites

- [ ] I have read [How we use GitHub Issues](https://github.com/OctopusDeploy/Issues/blob/main/docs/CONTRIBUTING.internal.md) for help deciding when and where it's appropriate to make an issue.
- [ ] I have considered informing or consulting the right people, according to the [ownership map](https://whimsical.com/ownership-map-NzbiD4HJyvhC9jNJNfS6TG).
- [ ] I have considered appropriate testing for my change.

### CODE REVIEW

File: source/Halibut.Tests/HalibutTimeoutsAndLimitsForTestsBuilder.cs
Comment (sburmanoctopus) [inline source/Halibut.Tests/HalibutTimeoutsAndLimitsForTestsBuilder.cs:52]: This feature is optional, but we are turning it on by default for our tests.

I think it makes more sense to run our tests the way they will be used by Octopus (i.e., we this off).

In saying that, it would be nice to see all the tests work with this on as well. Would it be overkill to double the number of tests and create different test cases with this on and off? 🤔 

Or do we just make a few choice tests with this on for now?

File: source/Halibut.Tests/Support/SerilogLoggerBuilder.cs
Comment (sburmanoctopus) [inline source/Halibut.Tests/Support/SerilogLoggerBuilder.cs:78]: Is this running locally, or on TeamCity?

I'm guessing it won't cause any issues running them in parallel?

File: source/Halibut.Tests/TraceLogFileLogger.cs
Comment (sburmanoctopus) [inline source/Halibut.Tests/TraceLogFileLogger.cs:58]: Just wondering, was this an issue locally or on TeamCity?

File: source/Halibut/HalibutRuntime.cs
Comment (sburmanoctopus) [inline source/Halibut/HalibutRuntime.cs:28]: Just to verify, the motivation behind this change was to make disposal async, right?

File: source/Halibut/Transport/SecureListener.cs
Comment (sburmanoctopus) [inline source/Halibut/Transport/SecureListener.cs:225]: I don't think I understand the subtlety here.

In the old one, we would wait until `cts.IsCancellationRequested || listener.Pending()` before calling accept.

Using the cancellation token removes the need for `cts.IsCancellationRequested`. But is waiting for `listener.Pending()` important? Or does this happen for free somehow with using the async versions of `AcceptTcpClientAsync`?

**Review by sburmanoctopus — COMMENTED** [review]
Looking good.

I'd like to understand the comment in `SecureListener` regarding `listener.Pending()` before approval though.

File: source/Halibut.Tests/HalibutTimeoutsAndLimitsForTestsBuilder.cs
Comment (LukeButters) [reply source/Halibut.Tests/HalibutTimeoutsAndLimitsForTestsBuilder.cs:52]: I think the direction would be to have this on so if anything we would want this on, not off.

Do we want a one off run with it off to check all that behaviour remains?

File: source/Halibut.Tests/TraceLogFileLogger.cs
Comment (LukeButters) [reply source/Halibut.Tests/TraceLogFileLogger.cs:58]: locally, my guess is crowd strike with slightly different timings (I had turbo on my laptop turned off, so I guess that increased the time files were opened for which I guess resulted in the conflict)

File: source/Halibut/HalibutRuntime.cs
Comment (LukeButters) [reply source/Halibut/HalibutRuntime.cs:28]: We needed to make accept be async (with a cancellation token)
Which meant that to finish the accept loop task we needed to await it which meant a IAsyncDisposable.
To do that meant we needed this to be a list of `IAsyncDisposable`
To do that and to preserve the only one thing can access this list at a time behaviour we needed some way of locking in both async and sync context (`lock` does not work with async)
To ensure that is done properly this type ensures, assuming you don't try really hard, that you can't have access to the list unless it is mutually exclusive access.

File: source/Halibut/Transport/SecureListener.cs
Comment (LukeButters) [reply source/Halibut/Transport/SecureListener.cs:225]: The whole wait until we know something is pending was done to avoid sitting in Accept at the same time as listener.stop() being called, because a bug did (or still does honestly I can't tell) where the call to listener.stop() would hang if AcceptTcpClient was in the middle of accepting a new connection.


The big change is to remove the SpinWait that `polls listener.pending()` because that is very expensive :(.
Instead we use the new (relative to when the old code was added) method that is `AcceptTcpClientAsync` with a cancellation token. When we shut down on linux the order is always:
1. Cancel the CTS
2. Wait for the loop to exit (which it does)
3. then and only then stop the listener (at this point we know we are NOT in accept and so can not hang).

File: source/Halibut.Tests/HalibutTimeoutsAndLimitsForTestsBuilder.cs
Comment (sburmanoctopus) [reply source/Halibut.Tests/HalibutTimeoutsAndLimitsForTestsBuilder.cs:52]: If the plan is to move it towards being the default behaviour, then this is fine.

**Review by sburmanoctopus — APPROVED** [review]

**Pull request merged**
